### PR TITLE
i#4508: Stolen register not loaded in dr_get_mcontext on AArch64

### PR DESCRIPTION
**Pull request by gregcawthorne**
i#4508 Dynamorio not loading stolen register value into mcontext on AArch64.

On AArch64 the value of the stolen register isn't loaded into the mcontext object for later instrumentation.

This is due to an ifdef only allowing this functionality on ARM (non AArch64), in core/lib/instrument.c, dr_get_mcontext_priv().

This causes errors when addresses need to be calculated based on the value of contents the stolen register.

Fix is a simple update of an #ifdef from #ifdef ARM to ifdef AARCHXX

### CODE REVIEW

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by AssadHashmi** [comment]
Added @gregcawthorne to Jenkins ACL for AArch64 pre-commit tests.

File: core/lib/instrument.c
Comment (derekbruening) [inline core/lib/instrument.c:6799]: (Putting commit title/description comment here to get an action item that can be marked Done and a separate thread.)

style nit 1: Title should have a colon after the issue number.

style nit 2: Description should have a final line that is either `Fixes #4508` or `Issue: #4508` (the former here).

style nit 3: suggestion: include the key `dr_get_mcontext` in the title, and remove the sort of implied "Dynamorio"  (should be "DynamoRIO", or "DR" for short) (it will also then fit in the github issue title); maybe sthg like `i#4508: Stolen register not loaded in dr_get_mcontext on AArch64`.

**Review by derekbruening — COMMENTED** [review]
Thank you for submitting the patch.  The code looks fine; just some nits/suggestions on the commit description.

**Comment by derekbruening** [comment]
Looks like the `client.drwrap-test-callconv` and `client.drwrap-test` tests fail on A64 with this patch with asserts which is surprising: can you reproduce that locally?

**Comment by gregcawthorne** [comment]
Hi.

Yes I am having the same problem.

The error is the same for both of them.

128: <Application /home/***/dynamorio/build/suite/tests/bin/client.drwrap-test-callconv (49428) DynamoRIO usage error : meta-instr faulted?  must set translation field and handle fault!>
128: <Usage error: meta-instr faulted?  must set translation field and handle fault! (/home/***/dynamorio/core/translate.c, line 909)

I will have to dig into this, this week.

Had been using in my drmemory work, but hadn't run the dynamoRIO tests.

**Comment by derekbruening** [comment]
PR #4491 very recently fixed an instance of that assert and maybe is not in your branch so it's worth an update to check, though I don't think the causes of the assert that it fixed (can be caused by multiple underlying issues) would show up on the drwrap tests.

File: core/lib/instrument.c
Comment (derekbruening) [inline core/lib/instrument.c:6803]: I happened to need this locally and it turns out there is a bug here: `dmc` can be NULL, so this should be:
```
    if (mc != NULL || TEST(DR_MC_INTEGER, dmc->flags)) {
```
I hit a crash with just this patch as-is due to this.

File: core/lib/instrument.c
Comment (derekbruening) [inline core/lib/instrument.c:6805]: Also, there seem to be 2 instances of the exact same `s/ARM/AARCHXX/` issue in `dr_set_mcontext` which could easily be part of this same PR.

Comment (derekbruening) [inline core/lib/instrument.c:6805]: It would also be nice to have a test that fails without each of the get and set fixes here (not to pile too much work on this...)

File: core/lib/instrument.c
Comment (gregcawthorne) [reply core/lib/instrument.c:6803]: Ah thanks,
I presume you mean:
`if (dmc != NULL || TEST(DR_MC_INTEGER, dmc->flags)) {`
?

File: core/lib/instrument.c
Comment (gregcawthorne) [reply core/lib/instrument.c:6805]: I should be able to create a test that mimic the original bug I found. Hopefully that will catch the absence of the other fix too, but will have to look into it.

File: core/lib/instrument.c
Comment (derekbruening) [reply core/lib/instrument.c:6803]: No, if `mc` is non-NULL, then there are no flags and everything is filled in and `dmc` is ignored and should be NULL.  Only if `mc` is NULL is `dmc` guaranteed to be non-NULL.

Comment (gregcawthorne) [reply core/lib/instrument.c:6803]: Ah okay that makes sense thanks!

**Comment by derekbruening** [comment]
Adding link to issue so it shows up there: #4508

**Comment by derekbruening** [comment]
My test for #4495 is hitting this as it wants to test the stolen register in a clean call so I'm slightly blocked on it.  @gregcawthorne -- when I put in the dr_set_mcontext fix too, the drwrap-test failures go away.

**Comment by gregcawthorne** [comment]
Ah okay that's good to know about the test failures going away, thanks.

Will confirm and push an updated version.

Will also look into making a test and see about getting that in tomorrow.

File: core/lib/instrument.c
Comment (gregcawthorne) [reply core/lib/instrument.c:6799]: For the description change, would you consider me force pushing a single commit with all the changes bad form. Or is it best to always do it incrementally and squash the commits before the merge.

Comment (derekbruening) [reply core/lib/instrument.c:6799]: Never force push.  The merge is always a squash with an edited description: edit it there.  Please remove all the star lines and create a single message at that time as instructed in the wiki.

**Comment by gregcawthorne** [comment]
Hi.

Update here.

Been putting together the test for this. At first I tried to make an api test, but now have tried to make a client test.

I have made one which calls dr_get_mcontext, dr_get_mcontext and opnd_compute_address, which seemed to pass with the fixes applied, but then I removed the fixes and it still passed.

I looked into it and its because these ifs return true from my test.

[https://github.com/DynamoRIO/dynamorio/blob/8f142e8909736b154d0ed09809c602b4497dac4a/core/lib/instrument.c#L6774
](url)
It's the dcontext->client_data->mcontext_in_dcontext part which return true here.


[https://github.com/DynamoRIO/dynamorio/blob/8f142e8909736b154d0ed09809c602b4497dac4a/core/lib/instrument.c#L6851](url)

This possibly explains how it hasn't been hit in testing previously.

 I am currently looking into how to make them return false for the sake of the test. One handicap seems to be how I can't seem to call into a lot of the nitty gritty methods and definitions directly from an external client, but that's by design probably. It should be possible to get this test to fail of course as DrMemory was and that is a client.

**Comment by derekbruening** [comment]
> I have made one which calls dr_get_mcontext, dr_get_mcontext and opnd_compute_address, which seemed to pass with the fixes applied, but then I removed the fixes and it still passed.
> 
> I looked into it and its because these ifs return true from my test.
> 
> [https://github.com/DynamoRIO/dynamorio/blob/8f142e8909736b154d0ed09809c602b4497dac4a/core/lib/instrument.c#L6774](url)

You mean your test calls `dr_get_mcontext` from a syscall event?  If you call it from a clean call from the cache you should hit the bug.  I hit it in every clean call from the cache I've added recently to tests: e.g., suite/tests/client-interface/stolen-reg.dll.c.

**Comment by derekbruening** [comment]
You could consider adding checks to suite/tests/client-interface/stolen-reg.{c,dll.c} rather than making a separate test if desired: that test is meant to detect stolen register bugs, after all, which this is an instance of, and I already hit it there when I checked the value in the clean callee.

**Comment by gregcawthorne** [comment]
Ah I didn't see those tests. They look comprehensive and not sure if my test would add much.

The only thing I was thinking is that isn't there an option to change the stolen reg from x28 to something else?

Perhaps the tests should find the stolen reg and then check based on that?

**Comment by derekbruening** [comment]
> Ah I didn't see those tests. They look comprehensive and not sure if my test would add much.

But stolen-reg.* is not comprehensive enough: it is not currently detecting this bug.

> The only thing I was thinking is that isn't there an option to change the stolen reg from x28 to something else?
> 
> Perhaps the tests should find the stolen reg and then check based on that?

Yes, that would be good, or at least a check for dr_get_stolen_reg() equaling x28 in the client main so we'll catch the test silently no longer testing what it should if we change the default stolen reg.

> But that should probably be a different commit.

If you could add it as part of adding the clean call `dr_get_mcontext` check that would be great.

**Comment by derekbruening** [comment]
> Ah I didn't see those tests. 

I just added the stolen-reg test last week which is why you didn't see it :)

**Comment by gregcawthorne** [comment]
Okay thanks. I will crack on!

File: suite/tests/client-interface/set-stolen-reg-mcontext.c
Comment (gregcawthorne) [inline suite/tests/client-interface/set-stolen-reg-mcontext.c:38]: I wasn't sure about checking for he stolen reg value in the app, as this fix is specifically about how the mcontext is maintained. For me I don't think the issue wast the app having a corrupt x28 register in its execution, but rather the client not knowing what the true value was.

File: suite/tests/client-interface/set-stolen-reg-mcontext.dll.c
Comment (gregcawthorne) [inline suite/tests/client-interface/set-stolen-reg-mcontext.dll.c:105]: Currently the fails with the fix in instrument.c, dr_set_mcontext reverted, due to a translation related error in recreate_app_state_from_ilist. However it passes with he fix and the recreate_app_state_from_ilist code path isn't hit with he fix. Currently trying to work out why this is the case.

File: suite/tests/client-interface/set-stolen-reg-mcontext.c
Comment (derekbruening) [reply suite/tests/client-interface/set-stolen-reg-mcontext.c:38]: The suggestion to possibly have the app check the value is for dr_set_mcontext where the client is changing the app value.  It is admittedly not a very normal thing for a real tool to do, but a cooperative app in a test is ok.  Another suggestion is to do what the existing cleancall tests do: consecutive clean calls where one does a dr_set_mcontext and the next does a dr_get_mcontext and looks for the new value.

Comment (gregcawthorne) [reply suite/tests/client-interface/set-stolen-reg-mcontext.c:38]: I have implemented the latter here. Can do a cooperative app-client test instead if that's best.

**Comment by abhinav92003** [comment]
@gregcawthorne This issue (#4508) is also causing `code_api|tool.drcachesim.delay-simple` test failures on AArch64 (#4571). We had to disable that test recently to unblock various other contributors. It'd be best to get the fix submitted soon, so that we can re-enable the test (not good to keep it disabled for long, as other bugs may creep in).

**Comment by gregcawthorne** [comment]
Have most of the tests passing now.

I seem to be having the same x86 test issue as #2433 is currently having, but only for the osx test.

Will look into the AArch64 Precommit failure tomorrow.

I tried to 'resolve conflicts' with he web UI and it said 'resolve commit and merge', so I hung fire as I wasn't sure if that was a merge into my dev branch or not.

**Comment by gregcawthorne** [comment]
I just resolved the merge conflicts, taking in the new changes from master, and all tests groups are failing now. I am sure were the cause of this is coming from currently.

**Comment by derekbruening** [comment]
https://github.com/DynamoRIO/dynamorio/issues/4370#issuecomment-738172341

**Comment by abhinav92003** [comment]
> I just resolved the merge conflicts, taking in the new changes from master, and all tests groups are failing now. I am sure were the cause of this is coming from currently.

Those failures were due to #4370, which is fixed now. Please update this branch to trigger the CI again.

**Comment by gregcawthorne** [comment]
run tests

**Comment by derekbruening** [comment]
I thought the plan was to just add a couple of checks to the existing stolen-reg test?

**Comment by gregcawthorne** [comment]
@derekbruening Will merge it into stolen reg then. Wasn't sure which way for definite, but was focusing on getting it fully passing first anyway.

**Comment by derekbruening** [comment]
If possible, please test offline to avoid so many pushes, to avoid stressing the CI resources: alternatively, please cancel the old ones.  These are all stacked up Appveyor which is globally serialized and is now tied up on just old commits for this PR for the next 3 to 4 hours: https://ci.appveyor.com/project/DynamoRIO/dynamorio/history

Please see https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#ci-resources-and-appveyor-global-serialization

**Comment by gregcawthorne** [comment]
Hi okay will do.

They were part formatting errors which I can test locally, part aarch32 errors which I am not currently set up to test as easily.

I naively hoped a new build would auto cancel the previous builds work.

Will try and cancel them now.

Thanks for the heads up

Greg

**Comment by derekbruening** [comment]
> @derekbruening Will merge it into stolen reg then. Wasn't sure which way for definite, but was focusing on getting it fully passing first anyway.

Sure.  Note that I did not look at the patch yet -- if the test is large and you prefer it separate that makes sense; I just remembered you saying it would fit into the existing test, and me also thinking it would be a natural fit there, but then seeing several commits here getting a separate test to build/work that might be avoided.

**Comment by gregcawthorne** [comment]
I have merged it into the original stolen-reg.dll.c test now anyway :)

**Comment by derekbruening** [comment]
> Hi okay will do. They were part formatting errors which I can test locally, part aarch32 errors which I am not currently set up to test as easily. I naively hoped a new build would auto cancel the previous builds work. Will try and cancel them now. Thanks for the heads up.

Travis would auto-cancel if a new commit was within a few seconds of a prior one.  Not sure about Github Actions -- @abhinav92003 do you recall seeing anything in the docs or notice anything there: it would be nice to auto-cancel old runs there.  I would say if there's a new commit at all before the old run is done, cancel it -- but not sure Github supports that.

**Comment by derekbruening** [comment]
Looks like it might be possible to use someone else's support for auto-cancel out of the box: https://github.community/t/is-it-possible-to-cancel-previous-jobs-automatically-when-the-pull-request-is-updated/16779

I'll file a feature request issue on it, if someone gets spare time (not sure how likely that is...)

**Comment by derekbruening** [comment]
> Looks like it might be possible to use someone else's support for auto-cancel out of the box: https://github.community/t/is-it-possible-to-cancel-previous-jobs-automatically-when-the-pull-request-is-updated/16779
> 
> I'll file a feature request issue on it, if someone gets spare time (not sure how likely that is...)

#4585

**Comment by derekbruening** [comment]
> add appveyor RDP login

Note that I canceled all the appveyor builds: just making sure you're not investigating a cancel or a flaky test: which test are you looking at on appveyor?

**Comment by gregcawthorne** [comment]
I just wanted to re-trigger the tests to get a 100% pass and I thought might as well add those lines you linked to in case it was useful in future.

**Comment by derekbruening** [comment]
Is this ready to review?  Github does not have a very good interface for "whose turn it is" so it is hard to tell: normally a "PTAL" comment helps when the code writer is finished, along with replying to all of the reviewer comments with "Done" or sthg.

**Comment by gregcawthorne** [comment]
Yes I believe could do with a review.

I still need to make the changes to the commit message you said, but that is to be done in the squash and merge I believe so didn't want to put 'Done'.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [inline suite/tests/client-interface/stolen-reg.dll.c:79]: Use `dr_fprintf(STDERR` in clients.

Comment (derekbruening) [inline suite/tests/client-interface/stolen-reg.dll.c:79]: suggest: Avoid duplication to avoid stale values when one changes: use `%s` and `__FUNCTION__`.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [inline suite/tests/client-interface/stolen-reg.dll.c:93]: I think it is better to have a DR_ASSERT or otherwise explicit check here that r10==test_value.  That makes the code easier to read (it's clear what we expect) and is generally better than an implied, hard-to-figure-out assertion through the "golden output" check of the output file.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [inline suite/tests/client-interface/stolen-reg.dll.c:101]: nit suggest: The client shouldn't really be aware of where DR is putting the stolen reg value, whether it's in TLS or somewhere in the private mcontext.  This is a black box test using public interfaces: so maybe s/save_stolen_reg_to_tls/change_stolen_reg_value/ and the other "read_and_restore_stolen_reg_value" or something.

**Review by derekbruening — COMMENTED** [review]
The basic content looks good but I have a number of suggestions that I think will improve the test.

**Comment by derekbruening** [comment]
We are again hitting this -- let's get it in.  I think I may just go make these suggested tweaks to get it merged.

File: .appveyor.yml
Comment (derekbruening) [reply .appveyor.yml:135]: Reverted.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:202]: Done.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:214]: Done.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:183]: Switched to a rarer sentinel value.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:210]: Done.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:176]: Removed.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:79]: Done.

Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:79]: Done.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:93]: Added the assert.  Just left the golden output though to expedite the changes.

File: suite/tests/client-interface/stolen-reg.dll.c
Comment (derekbruening) [reply suite/tests/client-interface/stolen-reg.dll.c:101]: Done.

**Comment by derekbruening** [comment]
Hmm, `0xdead` assembled in my local 32-bit arm cross-compile.

**Comment by derekbruening** [comment]
As mentioned, I made the requested changes myself to expedite this as we have been waiting for it too long and it is blocking a64 development.  Merging.

**Review by derekbruening — APPROVED** [review]

**Pull request merged**
